src/map.c

```diff
@@ -397,7 +403,27 @@ char *map_serialise_icemap(map_t *map, int *len)
 		memcpy(zf, pb, p-pb);
 		zf += p-pb;
 	}
-	memcpy(buf+buflen-8, "       \x00", 8);
+
+	// MapEnts
+	if (map->entities != NULL) {
+		memcpy(zf, "MapEnts", 7);
+		if (map->entities_size < 255) {
+			zf[7] = (uint8_t)map->entities_size;
+			zf += 8;
+		} else {
+			zf[7] = (char)'\xFF';
+			*(uint32_t *)&zf[8] = (uint32_t)map->entities_size;
+			zf += 12;
+		}
+		memcpy(zf, map->entities, map->entities_size);
+		zf += map->entities_size;
+	}
+
+	// MetaInf
+	// TODO:
+
+	// Terminator
+	memcpy(zf, "       \x00", 8);
 
 	//printf("derp!\n");
 	*len = buflen;
```

Comment:
I'd suggest `*len = zf - buf;` instead. As the MapEnts header can either be 8 or 12 bytes, this will produce extra garbage at the end of the file for a small MapEnts field.